internal/linker/link_node_modules.ts

```diff
@@ -1,764 +1,234 @@
-// of truth for the checked-in file.
-const {runfiles: _defaultRunfiles, _BAZEL_OUT_REGEX}:
-    typeof import('@bazel/runfiles') = require('../runfiles/index.js')
-import {Runfiles} from '@bazel/runfiles';
-
 // Run Bazel with --define=VERBOSE_LOGS=1 to enable this logging
 const VERBOSE_LOGS = !!process.env['VERBOSE_LOGS'];
 
-function log_verbose(...m: string[]) {
-  if (VERBOSE_LOGS) console.error('[link_node_modules.js]', ...m);
-}
-
-function log_error(error: Error) {
-  console.error('[link_node_modules.js] An error has been reported:', error, error.stack);
+function log(...m: string[]): void {
+  if (VERBOSE_LOGS) {
+    console.error(loggingPrefix, ...m);
+  }
 }
 
-/**
- * Create a new directory and any necessary subdirectories
- * if they do not exist.
- */
-async function mkdirp(p: string) {
-  if (p && !await exists(p)) {
-    await mkdirp(path.dirname(p));
-    log_verbose(`creating directory ${p} in ${process.cwd()}`);
-    try {
-      await fs.promises.mkdir(p);
-    } catch (e) {
-      if (e.code !== 'EEXIST') {
-        // can happen if path being created exists via a symlink
-        throw e;
-      }
-    }
+function fatal(context: string, errors: Error[]): never {
+  console.error(loggingPrefix, 'Error(s) were reported.');
+  console.error(loggingPrefix, 'Context:', context);
+  for (const error of errors) {
+    console.error(error);
   }
+  console.error(loggingPrefix, 'Exiting');
+  process.exit(1);
 }
 
+/** Runfiles directory path. e.g. `/__output_base__/execroot/com_canva_canva/bazel-out/___/bin/___/___.runfiles/web_node_modules/node_modules` */
+export type RunfilesDirPath = string & { __RunfilesDirPath: never };
+/** Mount point for `node_modules` relative to workspace root. e.g. `web`, `web/tools/cloudflare` */
+export type RelativeMountPath = string & { __RelativeMountPath: never };
+/** A workspace name. e.g. `com_canva_canva`, `rules_nodejs`, `web_node_modules` */
+export type WorkspaceName = string & { __RepositoryName: never };
+
 /**
- * Gets the `lstat` results for a given path. Returns `null` if the path
- * does not exist on disk.
+ * Manifest defining `node_modules` mounts that need to be created, and for which workspace.
+ * File defined in `@rules_nodejs//internal/linker/link_node_modules.bzl`
  */
-async function gracefulLstat(path: string): Promise<fs.Stats|null> {
-  try {
-    return await fs.promises.lstat(path);
-  } catch (e) {
-    if (e.code === 'ENOENT') {
-      return null;
-    }
-    throw e;
-  }
-}
+type ModulesManifest = {
+  roots: Record<RelativeMountPath, WorkspaceName>,
+  workspace: WorkspaceName,
+};
+
+/** Path for symlink file (aka link name). */
+type SymlinkFilePath = string & { __SymlinkFilePath: never };
+/** Path for symlink target. */
+type SymlinkTargetPath = string & { __SymlinkTargetPath: never };
+/** Describes potential symlinks. */
+type ProposedSymlink = [filePath: SymlinkFilePath, targetPath: SymlinkTargetPath];
```

Comment:
nit: object would be clearer
```suggestion
type ProposedSymlink = { fromPath: SymlinkFilePath, targetPath: SymlinkTargetPath };
```